app/components/form/fields/TlsCertsField.tsx

```diff
@@ -0,0 +1,172 @@
+import { useForm } from 'app/hooks'
+
+export function TlsCertsField({ control }: { control: Control<SiloCreateInput> }) {
+  const [showAddCert, setShowAddCert] = useState(false)
+
+  const {
+    field: { value: items, onChange },
+  } = useController({ control, name: 'tlsCertificates' })
+
+  return (
+    <>
+      <div className="max-w-lg">
+        <FieldLabel id="tls-certificates-label" className="mb-3">
+          TLS Certificates
+        </FieldLabel>
+        {!!items.length && (
+          <MiniTable.Table className="mb-4">
+            <MiniTable.Header>
+              <MiniTable.HeadCell>Name</MiniTable.HeadCell>
+              {/* For remove button */}
+              <MiniTable.HeadCell className="w-12" />
+            </MiniTable.Header>
+            <MiniTable.Body>
+              {items.map((item, index) => (
+                <MiniTable.Row
+                  tabIndex={0}
+                  aria-rowindex={index + 1}
+                  aria-label={`Name: ${item.name}, Description: ${item.description}`}
+                  key={item.name}
+                >
+                  <MiniTable.Cell>{item.name}</MiniTable.Cell>
+                  <MiniTable.Cell>
+                    <button
+                      onClick={() => onChange(items.filter((i) => i.name !== item.name))}
+                    >
+                      <Error16Icon title={`remove ${item.name}`} />
+                    </button>
+                  </MiniTable.Cell>
+                </MiniTable.Row>
+              ))}
+            </MiniTable.Body>
+          </MiniTable.Table>
+        )}
+
+        <Button size="sm" onClick={() => setShowAddCert(true)}>
+          Add TLS certificate
+        </Button>
+      </div>
+
+      {showAddCert && (
+        <AddCertModal
+          onDismiss={() => setShowAddCert(false)}
+          onSubmit={async (values) => {
+            const certCreate: (typeof items)[number] = {
+              ...values,
+              cert: await values.cert.text(),
+              key: await values.key.text(),
+            }
+            onChange([...items, certCreate])
+            setShowAddCert(false)
+          }}
+          allNames={items.map((item) => item.name)}
+        />
+      )}
+    </>
+  )
+}
+
+export type TlsCertificate = Omit<
+  SiloCreateInput['tlsCertificates'][number],
+  'key' | 'cert'
+> & {
+  key: File
+  cert: File
+}
+
+const defaultValues: Partial<TlsCertificate> = {
+  description: '',
+  name: '',
+  service: 'external_api',
+}
+
+function UniqueNameField<
+  TFieldValues extends FieldValues,
+  TName extends FieldPath<TFieldValues>
+>({
+  required = true,
+  name,
+  label = capitalize(name),
+  allNames,
+  ...textFieldProps
+}: Omit<TextFieldProps<TFieldValues, TName>, 'validate'> & {
+  label?: string
+  allNames: string[]
+}) {
+  return (
+    <TextField
+      validate={(name) =>
+        allNames.includes(name) ? 'Certificate with this name already exists' : true
```

Comment:
I did attempt to get this going in the `onSubmit` of `AddCertModal` but for whatever reason nothing I put in there was firing. Guessing `react-hook-form` is intercepting the default `onSubmit`.

But the benefit of this approach is we don't need to mess around with `setError` or anything like that. I did consider memoizing `allNames` but figured it might be overboard for the number of certs we'll have here.